xmppserver/src/main/java/org/jivesoftware/openfire/pubsub/PubSubPersistenceProviderManager.java

```diff
@@ -58,7 +58,16 @@ public PubSubPersistenceProvider getProvider() {
 
     private void initProvider()
     {
-        final Class clazz = PROVIDER.getValue();
+        Class clazz = PROVIDER.getValue();
+        if ( clazz == null ) {
+            if ( ClusterManager.isClusteringEnabled() ) {
+                Log.debug("Clustering is enabled. Falling back to non-cached provider");
+                clazz = DefaultPubSubPersistenceProvider.class;
+            } else {
+                clazz = CachingPubsubPersistenceProvider.class;
+            }
+        }
+
         // Check if we need to reset the provider class
         if (provider == null || !clazz.equals(provider.getClass()) ) {
```

Comment:
Not sure this code branch is necessary, as the non-dynamic provider can't change, so no need to shut down the old provider.

Or make it dynamic, and add this method as an onChange method.